sklearn/ensemble/_hist_gradient_boosting/splitting.pyx

```diff
@@ -716,6 +776,197 @@ cdef class Splitter:
                 split_info.sum_gradient_right, split_info.sum_hessian_right,
                 lower_bound, upper_bound, self.l2_regularization)
 
+    @cython.initializedcheck(False)
+    cdef void _find_best_bin_to_split_category(
+            self,
+            unsigned int feature_idx,
+            unsigned char has_missing_values,
+            const hist_struct [:, ::1] histograms,  # IN
+            unsigned int n_samples,
+            Y_DTYPE_C sum_gradients,
+            Y_DTYPE_C sum_hessians,
+            Y_DTYPE_C value,
+            char monotonic_cst,
+            Y_DTYPE_C lower_bound,
+            Y_DTYPE_C upper_bound,
+            split_info_struct * split_info) nogil:  # OUT
+        """Find best split for categorical features. """
+
+        cdef:
+            unsigned int bin_idx
+            unsigned int end = self.n_bins_non_missing[feature_idx]
+            unsigned int missing_values_bin_idx = self.missing_values_bin_idx
+            categorical_info * cat_sorted_infos
+            unsigned int sorted_idx
+            unsigned int n_used_bin = 0
+            const hist_struct[::1] feature_hist = histograms[feature_idx, :]
+            Y_DTYPE_C sum_gradients_bin
+            Y_DTYPE_C sum_hessians_bin
+            Y_DTYPE_C loss_current_node
+            # Reduces the effect of noises in categorical features,
+            # especially for categoires with few data
+            # TODO: Make this user adjustable?
+            unsigned int MIN_CAT_SUPPORT = 10
+            # Used for find best split
+            unsigned int MAX_CAT_THRESHOLD = 32
```

Comment:
Regarding `MAX_CAT_THRESHOLD` and scanning in both directions: we had agreed not to do this IIRC?

I still don't understand why this is needed. If there's a strong incentive to keep it, let's please document it, otherwise let's remove: we don't need some extra complexity that we don't understand yet.

Reply:
From memory, when I removed this, the overall test performance was less than lightgbm.

I guess I can remove this again just to further simplify things and knowing that the splits found may be different from lightgbm.

Reply:
Removed this and replaced with a full scan in one direction.